### ISSUE TITLE
tournament handle gsub is messed and handle should be editable

### ISSUE THREAD

**Issue by kevinhughes27** [kevinhughes27/tournament-app#11]
also missing validations.

**Comment by theoandtheb** [comment]
I noticed that, too.  I had been working on a client-side solution to this, but was swept up quickly by other tasks.  Would you prefer a refined server-side solution?  Both could work, in theory.  We could use an .on("change") with a regex in the handler to check for invalid chars and alert the user if the input will cause trouble, then expand set_handle to facilitate percent-encoding.

Of course, as I write this, I notice you have already made changes to handle characters outside the  RFC 3986 spectrum.  I'll go worry about something else now.

**Comment by kevinhughes27** [comment]
yeah the on-change was what I was planning on doing as well. As soon as they have made it custom once then we can kill the on change. (this is how shopify does product handles in our urls). Then we can just validate the handle on save.

**Comment by theoandtheb** [comment]
That sounds good.  I'll see if I can find an efficient way to test this on
the client-side and at least get something working on every change to that
field (new and edit).  Though a little hacky, since the field exists in a
partial, I suppose I could place the script at the end of the new view,
rather than within the tournament coffeescript.

Personally, I think the validation reminder should appear both when
creating and editing the handler, as it is quite possible a user would
create a tournament with a valid handler then edit it to an invalid one.
The best solution, I believe, would be an ajax call which returns the
handle as it would be generated on change.  That may be a lot of work for a
minor feature, but this would ensure the user is satisfied with the
outcome  before completing the process.  Considering how the handle will
likely be used for promotion, it seems worthy of consideration.

**Comment by kevinhughes27** [comment]
what I was thinking was:

Client side: 
as tournament title is changed update the handle unless the handle already doesn't match the handlized title. This happens on new and edit tournament until the user types a custom handle

Server Side:
validate that handle has only valid chars
validate uniqueness of handle

Client side validations are nice but hard to maintain and they can't be relied on

**Comment by theoandtheb** [comment]
You lost me with this sentence:

"as tournament title is changed update the handle unless the handle already
doesn't match the handlized title"

The first part I get, but I am unclear as to what you are describing with
the unless condition.

Based on my limited testing, it seems as if the set_handle method is now
truncating invalid characters, which seems fine to me provided the user
understands this will take place.  Client-side validation is a pain, I
agree, especially when you consider just how many possible characters are
invalid.

You didn't comment on it specifically, but what do you think of the idea I
proposed of using an ajax call to return the value of set_handle based on
the value of the name input field on change?  This would allow the user to
adjust the title to conform to how our method generates the handle,
effectively allowing the user to self-validate before submitting the form.
Anytime we change the underlying method, the "validation" will
automatically adjust.

**Comment by kevinhughes27** [comment]
hmm its a hard thing to describe. So its not so much client side validations that I want but more client side auto-filling. So rather than an ajax call we can just have an on change callback that updates the handle field whenever the title is changed on the form.

In this callback that autofills that form field we can check if the handle has deviated from what we would automatically suggest given the title. If this is the case then we stop updating it along with the title.

Then on the server we simply check that the handle that was sent to use only contains valid chars.

Does that make sense?

**Comment by theoandtheb** [comment]
So, what you are suggesting is logic on the client-side that will mimic the
handle generation performed on the server-side?  If not, perhaps I should
leave this to you and stop worrying about it, as I am still not quite sure
what it is you have in mind.  If that is the case, I'm struggling to see
how this is an improvement upon using ajax to send the value as entered and
have the server respond with the handle it is designed to return.  Should
the method be changed for any reason, a client-side auto-fill logic would
need to change as well.

If there is a fundamental problem with the ajax-based proposition I'm all
ears, but it seems like both the simplest and DRYest approach of the two.

**Comment by kevinhughes27** [comment]
ajax is a bit excessive for this imo. Yeah why don't I mock it up and then hopefully it makes sense. I might be missing something that makes this hard than I think it should be. 

If you want to work on this our regex still needs to be fixed since its not replacing spaces with dashs anymore. Its needs 2 steps - replace invalid chars like `'` with nothing and spaces with dashes

**Comment by theoandtheb** [comment]
Sure.  I'll see what I can piece together using whatever logic is in place
for generating handles.  If this works, I can then work with you or anyone
else who helped generate the existing set_handle method.  I haven't done
much with friendly_id in the past, but am willing to get up to speed and
help revise the regex.

**Comment by theoandtheb** [comment]
I'll get to work on the regex later today, but I now have a working on
change that runs the tournament's name field's value through the set_handle
method and replaces the text within a div under the input line with a
prompt showing the URL of the app controller (root/{set_handle output}).  I
can change this to some other value, though I suspect the plan is to feed
all inbound traffic for a given tournament through the App Controller.

**Comment by kevinhughes27** [comment]
that is the plan! thats exactly it. 

For products in Shopify we show it like this:
![image](https://cloud.githubusercontent.com/assets/1965489/8044897/d953e6f4-0dfd-11e5-9e65-b27f84bef627.png)
